Fix issue 3155 by running `scancode-reindex-licenses` subcommand instead of using `--reindex-licenses` flag

abhi-kr-2100 opened the pull request:
<!-- Delete Template sections if unneccesary -->
<!-- Add issue number here (We encourage you to create the Issue First) -->
<!-- You can also link the issue in Commit Messages -->

Fixes #3155 

This PR only modifies the Dockerfile. In the erroneous Dockerfile, `./scancode --reindex-licenses` was used to
1. Configure venv
2. Reindex licenses

Since the `--reindex-licenses` flag has been removed, this led to an error. The fix is to use the new subcommand `scancode-reindex-licenses` to do the reindexing. Configuration is done prior to the reindexing using the `./configure` script.

<!--
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!
* [x] - Checked Box
* [ ] - Unchecked Box
-->

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->
Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

abhi-kr-2100 committed:
Add `scancode-reindex-licenses` script

This script provides a CLI entry point for the `scancode-reindex-
licenses` subcommand introduced by
5361052117189489703d436a68e5f364f9a39c0d.

Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

abhi-kr-2100 committed:
Use `scancode-reindex-licenses` to generate index

The `--reindex-licenses` flag has been removed from the main
`scancode` executable into the separate subcommand
`scancode-reindex-licenses`.

Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

abhi-kr-2100 committed:
Use the subcommand `scancode-reindex-licenses`

As reindexing licenses is an exceptional operation, a separate
script is not preferable. Hence, the script has been removed.

In the earlier Dockerfile, `./scancode --reindex-licenses`
did two jobs:
1. The initial configuration
2. Reindexing of licenses

In the new Dockerfile, `./configure` is first run to
configure the venv. This is followed by a run of
`scancode-reindex-licenses`.

Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

abhi-kr-2100 committed:
Remove extra edits caused by autoformatting

Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

abhi-kr-2100 committed:
Add CI test for building and scanning using Docker

Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

s01ipsist commented on Dockerfile:
```diff
@@ -37,9 +37,11 @@ WORKDIR /scancode-toolkit
+RUN ./configure
+
+# Run scancode-reindex-licenses to build the base license index
+RUN ./venv/bin/scancode-reindex-licenses
```
Dockerfile best practice is to combine successive `RUN` commands so as to minimise image layers.

```
RUN ./configure && \
    ./venv/bin/scancode-reindex-licenses
```
Use comments as you prefer

abhi-kr-2100 replied:
Thank for taking the time to review.

I've combined the two into a single RUN instruction.

s01ipsist commented on azure-pipelines.yml:
```diff
@@ -7,6 +7,27 @@
+          'scancode-toolkit'
+          'scancode-toolkit:$(Build.BuildId)'
+    - script: |
+        docker scan scancode-toolkit:$(Build.BuildId)
```
I understand the recommendation was to use the built image to run a scan.
https://scancode-toolkit.readthedocs.io/en/latest/getting-started/install.html#run-using-docker
Has some sample commands

abhi-kr-2100 replied:
Thanks a lot for pointing me in the right direction; I had gotten myself on a completely wrong track.

I've updated the `script` key with a sample run of `scancode`.

abhi-kr-2100 committed:
Combine configuration and reindexing into one RUN instruction

Dockerfile best practice is to combine successive RUN
commands so as to minimise image layers.

Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

abhi-kr-2100 committed:
Add CI job to test build and run with Docker

The test should ensure that scancode-toolkit is capable of
being built successfully using Docker. Additionally, the built
script should run successfully.

Signed-off-by: Abhishek Kumar <abhi.kr.2100@gmail.com>

s01ipsist reviewed (commented):
Looks ok to me @abhi-kr-2100 but we need a project maintainer to review

abhi-kr-2100 commented:
Thanks for taking the time to review this @s01ipsist.

Let me gently ping @pombredanne.

pombredanne reviewed (approved):
LGTM! thank you ++
Merging now

pombredanne commented:
@abhi-kr-2100 Thanks!

pombredanne commented:
@abhi-kr-2100 I did not pay attention to the details, but your changes have disabled the Azure pipelines entirely. I am going to revert this urgently

abhi-kr-2100 commented:
@pombredanne Oops. Please do the needful.
